Support storing TSQL view definition in Babelfish catalog

HarshLunagariya opened the pull request:
Previously, TSQL view definitions were not supported because internally
definitions didn't get stored in the catalog, they get de-parsed at runtime in
pg dialect from tuple and query tree data.

This commit adds a Babelfish catalog babelfish_object_def for storing object
definitions along with db, schema, object name, flags of settings like ansi nulls,
quoted identifier. It implements pltsql_store_view_definition_hook function to 
store TSQL view definition along with flags into catalog. pltsql_drop_view_definition
function is implemented to drop the TSQL definition from catalog when
object_access_hook is invoked on a drop object event for TSQL
view.

This commit blocks DDLs for TSQL view objects which has entry in catalog from
PG dialect by introducing enable_create_alter_view_from_pg GUC which will
block DDL statements from PG dialect. enable_create_alter_view_from_pg GUC's
value will be false by default i.e. DDL's will be blocked. Currently it is only blocking
DDLs for view objects from PG endpoint. It is required to set this GUC to true in
upgrade script and while creating any internal views.

It implements information_schema_tsql.views such that it returns data related to
every view which is accessible to current user.

Task: BABEL-3135,2889
Signed-off-by: Harsh Lunagariya [lunharsh@amazon.com](mailto:lunharsh@amazon.com)

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

HarshLunagariya committed:
Support storing TSQL view definition in Babelfish catalog

Previously, TSQL view definitions were not supported because internally
definitions didn't get stored in the catalog, they get de-parsed at runtime in
pg dialect from tuple and query tree data.

This commit adds a babelfish catalog bbf_object_def for storing object
definitions along with db, schema and object name.It implements
pltsql_store_view_definition_hook function to store TSQL view definition into
babelfish catalog. pltsql_drop_view_definition function is implemented
to drop the TSQL definition from catalog when object_access_hook is
invoked on a drop object event for TSQL view.

Task: BABEL-3135
Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Temporarily use 'amazon-aurora:jira-babel-3135' for building engine

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

2jungkook commented on contrib/babelfishpg_tsql/sql/object_definition_tsql.sql:
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE sys.babelfish_view_def (
```
I would suggest this table to have only oid (pkey) and view definition columns. Otherwise, we need additional special handlings for `ALTER DATABASE` and `ALTER SCHEMA`.

HarshLunagariya replied:
The problem with having oid as primary key is that oid might get changed during Major Version Upgrade/Pg dump-restore, in that case we've to figure out the oid and update the catalog which will be more complex. So it is better to have combined key of db, schema and object name and handle the cases for `ALTER DATABASE` and `ALTER SCHEMA`.

2jungkook replied:
Okay, that makes sense. I'm not sure whether all those functionalities are in the current BBF. If not, let's add at least test cases to notify future developers :) FYI, object name may also be changed by `sp_rename`.

If we are storing those names, I'd suggest to avoid using pg-only types because it can cause different result for comparisons. I guess T-SQL has their own types for db/schema/object names.

2jungkook commented on contrib/babelfishpg_tsql/sql/object_definition_tsql.sql:
```diff
@@ -0,0 +1,8 @@
+	db_name NAME NOT NULL,
+	schema_name NAME NOT NULL,
+	object_name NAME NOT NULL,
+	definition TEXT NOT NULL COLLATE "C",
```
Let's use `nvarchar(max)` to align with the definition of `sys.sql_modules` (https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/relational-databases/system-catalog-views/sys-all-sql-modules-transact-sql.md). Btw, Is there any reason to use C collation here?

KushaalShroff replied:
Even if we use nvarchar(max) we might need to cast it to different string types of different lengths while using it, since sys.sql_modules is not the only system view which stores the definitions

2jungkook replied:
I'm afraid of using pg-only types (NAME, TEXT) in BBF catalogs because they behave differently, e.g., ANSI_PADDINGS. And would you point me to some examples where different system views use different data types for the object definitions or names?

HarshLunagariya replied:
Fixed. Used TSQL types in catalog definition.

2jungkook commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--1.2.0--1.3.0.sql:
```diff
@@ -914,5 +914,14 @@ AS $$
+	PRIMARY KEY(db_name, schema_name, object_name)
+);
+GRANT SELECT ON sys.babelfish_view_def TO PUBLIC;
+
```
What is our policy for views created before this change? Are we going to keep those entries with null view definition or just to ignore them?

KushaalShroff replied:
For now, there wont be any entires for views prior to this change. We can ask the User to either drop and re-create the view they are blocked on or we can provide some other way to insert their view definition into this table (The first workaround seems to be the better one)

2jungkook commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -1454,3 +1467,125 @@ modify_insert_stmt(InsertStmt *stmt, Oid relid)
+ * Stores view object's TSQL definition to bbf_view_def catalog
+ */
+static void
+pltsql_store_view_definition(const char *queryString, ObjectAddress address)
```
There can be two corner cases:
1. Creating a view in PG connection with sql_dialect = T-SQL
2. `ALTER VIEW ... ALTER COLUMN` in PG connection

HarshLunagariya replied:
1. First case is handled correctly by current implementation.
2. For `ALTER VIEW ... ALTER COLUMN`, We shouldn't allow user to alter TSQL views from pg endpoint. For that I'll add the checks in engine.

Will also add required test cases.

2jungkook replied:
If we execute T-SQL in a PG connection, it won't go through ANTLR parser so the result can be different, e.g., `select 'value' = '1';`. Also I just realized that we modify query strings in both ANTLR and parse_analyze stages. Could you check whether we store the original query or modified one? For example,
- `select * from db..object` is changed to `select * from db.dbo.object`
- `select * from dbo.object` is changed to `select * from master_dbo.object`

HarshLunagariya replied:
Currently, We've blocked DDL from pg endpoint using GUC.

surendravishnoi commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--1.2.0--1.3.0.sql:
```diff
@@ -914,5 +914,14 @@ AS $$
 $$
 LANGUAGE SQL IMMUTABLE PARALLEL RESTRICTED;
```
this file is deleted in git farm

HarshLunagariya committed:
Block ALTER/REPLACE statememnts from PG dialect on TSQL views

Commit changes:
- Blocked the ALTER/REPLACE VIEW statement execution on the TSQL which has entry
  in babelfish_view_def catalog.
- Implemented check_is_tsql_view() function to check if TSQL view has
  entry in babelfish_view_def catalog.
- Changed definition of babelfish_view_def catalog
- Implemented clean_up_bbf_view_def() function to cleanup respective
  entries when when drop babelfish db.
- Truncate the definition to 8000 character length while storing it in
  catalog.
- Implemented get_dbid_from_physical_schema_name function to get logical
  dbid from physical schema name by looking up into
  babelfish_namespace_ext catalog

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
Temporarily use jira-babel-3135 engine branch for upgrad tests

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Support TSQL's Information_schema.views

Implemented the information_schema_tsql.views view which returns one row for
views that can be accessed by the current user in the current database.

Added test cases for information_schema.views in JDBC test framework.

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Move code for blocking ALTER/REPLACE statements for TSQL views from pg endpoint

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

sangilsong commented on contrib/babelfishpg_tsql/sql/object_definition_tsql.sql:
```diff
@@ -0,0 +1,8 @@
+	dbid SMALLINT NOT NULL,
+	schema_name NAME NOT NULL,
+	object_name NAME NOT NULL,
+	definition TEXT NOT NULL COLLATE "C",
```
Don't we need to capture session property (i.e. ANSI_NULLS)? In my short experiment, select on SQL view follows the setting at when it is created.

```
1> create table tt1(a varchar(10));
2> go
1> insert into tt1 values (null);
2> go
1> set ansi_nulls off;
2> go
1> select * from tt1 where a = null;
2> go
a
----------
NULL

1> create view v11 as select * from tt1 where a = null;
2> go
1> select * from v11;
2> go
a
----------
NULL

-- v11 behaves like ansi_nulls=off
1> set ansi_nulls on;
2> go
1> select * from v11;
2> go
a
----------
NULL

(1 rows affected)
1> select * from tt1 where a = null;
2> go
a
----------

(0 rows affected)
```

HarshLunagariya replied:
Yes, we should add session properties like ANSI_NULLS, QUOTED_IDENTIFIER. I will address this in upcoming commits.

sangilsong commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -1290,6 +1300,9 @@ bbf_object_access_hook(ObjectAccessType access, Oid classId, Oid objectId, int s
 		(*prev_object_access_hook) (access, classId, objectId, subId, arg);
 
+	if (access == OAT_DROP && classId == RelationRelationId)
+		pltsql_drop_view_definition(objectId);
```
is this hook called when base object which a T-SQL view depends on is dropped cascade?

HarshLunagariya replied:
yes. it will be called.
Next question might be then why `clean_up_bbf_view_def()` is needed?
Reason: 
- When we drop logical db in babelfish, `drop_bbf_db()` function gets executed. 
- In that function we generate subcommands to drop child objects.
- At that time this hook is called but in that case we're unable to lookup babelfish_namespace catalog to get logical schema name. So need to explicitly clean the catalog in case of `drop_bbf_db()`

HarshLunagariya committed:
Change datatype to sysname from name in catalog definition

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

xhfanhe commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -774,6 +830,109 @@ get_authid_user_ext_db_users(const char *db_name)
+		if (oldtup)
+			elog(ERROR,
+				 "multiple babelfish_view_def entries for object %u",
+				 relid);
```
Under what circumstances will there be multiple entries for single object?

HarshLunagariya replied:
Yes, It's redundant because we're defining primary key on dbid, schema name and object name. I will address it in next commits.

xhfanhe commented on .github/workflows/minor-version-upgrade.yml:
```diff
@@ -10,9 +10,9 @@ on:
+      ENGINE_VER_FROM: jira-babel-3135
       EXTENSION_VER_FROM: BABEL_2_0_STABLE
-      ENGINE_VER_TO: BABEL_2_X_DEV__PG_14_2
+      ENGINE_VER_TO: jira-babel-3135
```
Please make sure to reset yml changes before final merge

xhfanhe commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -774,6 +830,109 @@ get_authid_user_ext_db_users(const char *db_name)
+	if (HeapTupleIsValid(oldtup))
+		is_tsql_view = true;
+
+	heap_freetuple(oldtup);
```
Do we need to free scantup here as well?

HarshLunagariya replied:
Documentation for `systable_getnext(SysScanDesc sysscan)` function goes like:
`Note that returned tuple is a reference to data in a disk buffer; it must not be modified, and should be presumed inaccessible after next getnext() or endscan() call.` 
So, We don't need to free scantup.

xhfanhe reviewed (changes requested):
Left some comments, please take a look.
Also, with this change, it would be safer to add more rules in metadata integrity check, in which way we can check if babelfish_view_def catalog is working correctly. Please consider filing a separate Jira for that.

HarshLunagariya committed:
Add Additional columns for session properties and object properties columns and use TSQL datatypes in catalogs

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3135

shalinilohia50 commented on test/JDBC/input/bbf_view_def.mix:
```diff
@@ -0,0 +1,168 @@
+-- tsql
+create table vd_t1 (c1 int primary key);
```
Please drop this table at the end. It leads to one failure in JDBC.

HarshLunagariya replied:
Done.

HarshLunagariya committed:
Fix test failure due to un-dropped objects

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

surendravishnoi commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -401,4 +401,41 @@ CREATE VIEW information_schema_tsql.table_constraints AS
+			CAST(
+				CASE WHEN LENGTH(vd.definition) <= 4000
+					THEN vd.definition
+					ELSE NULL END
```
do we want to return NUL or truncated value?

HarshLunagariya replied:
Yes, We want to return NULL and it is expected . See [Documentation](https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/relational-databases/system-information-schema-views/views-transact-sql.md)

surendravishnoi commented on contrib/babelfishpg_tsql/sql/object_definition_tsql.sql:
```diff
@@ -0,0 +1,14 @@
+CREATE TABLE sys.babelfish_view_def (
```
not for this change but I think we should device a way to create such tables in binary itself. That will save a lot of trouble for upgrades in future

surendravishnoi commented on contrib/babelfishpg_tsql/src/catalog.h:
```diff
@@ -89,6 +89,7 @@ extern const char *get_one_user_db_name(void);
 #define NAMESPACE_EXT_TABLE_NAME "babelfish_namespace_ext"
 #define NAMESAPCE_EXT_PK_NAME "babelfish_namespace_ext_pkey"
 #define Anum_namespace_ext_namespace 1
+#define Anum_namespace_ext_dbid 2
```
please make sure that we never used this id in previous releases

HarshLunagariya replied:
Double Checked. We haven't used it yet.

surendravishnoi commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -1454,3 +1468,145 @@ modify_insert_stmt(InsertStmt *stmt, Oid relid)
+ * Drops view object's TSQL definition from bbf_view_def catalog
+ */
+static void
+pltsql_drop_view_definition(Oid objectId)
```
it can share code with it_tsql view

surendravishnoi commented on contrib/babelfishpg_tsql/src/pl_handler.c:
```diff
@@ -2045,6 +2045,81 @@ static void bbf_ProcessUtility(PlannedStmt *pstmt,
+				if (vstmt->replace && check_is_tsql_view(relid))
+				{
+					ereport(ERROR,
+							(errcode(ERRCODE_INTERNAL_ERROR),
```
if it is triggered by user action, it should not be an internal error. Rather not supported or something like that

HarshLunagariya replied:
Using ERRCODE_FEATURE_NOT_SUPPORTED now. Fixed.

Deepesh125 commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -413,6 +424,51 @@ get_logical_schema_name(const char *physical_schema_name, bool missingOk)
+	if (get_namespace_oid(physical_schema_name, false) == InvalidOid)
+		return NULL;
+
+	rel = table_open(namespace_ext_oid, AccessShareLock);
```
why access share lock here? shouldn't we use row exclusive lock?

HarshLunagariya replied:
No need for row exclusive lock as we use this function only for getting dbid using namespace_ext_oid relation. If you're considering DROP DB/SCHEMA case then we must be holding RowExclusiveLock while dropping.

HarshLunagariya committed:
Refactor pltsql_drop_view_definition() and check_is_tsql() and change in error messages

2jungkook commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -774,6 +830,118 @@ get_authid_user_ext_db_users(const char *db_name)
+}
+
+bool
+check_is_tsql_view(Oid relid)
```
Could you add a comment explaining that this function will return false if a T-SQL view was created before v2.1?

2jungkook commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -1454,3 +1468,147 @@ modify_insert_stmt(InsertStmt *stmt, Oid relid)
+	tuple = heap_form_tuple(bbf_view_def_rel_dsc,
+							new_record, new_record_nulls);
+
+	CatalogTupleInsert(bbf_view_def_rel, tuple);
```
Can this support rollback? What if an exception happens after pltsql_store_view_definition()?

HarshLunagariya replied:
Yes, it supports rollback, I will add test case for the same. And I checked using inserting an exception after pltsql_store_view_definition(), it is rolling back as expected.

2jungkook commented on test/JDBC/expected/bbf_view_def.out:
```diff
@@ -0,0 +1,308 @@
+~~ERROR (Code: 33557097)~~
+
+~~ERROR (Message: 'ALTER VIEW' is not currently supported in Babelfish)~~
+
```
Let's add a select statement here as well because once we support alter view then the result of the select statement should be changed.

HarshLunagariya committed:
Implemented GUC enable_dml_ddl_from_pgendpoint

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
Move upgrade changes from 2.0-2.1 to 2.1-2.2 script

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
[Temporary Change] Update workflow config to run mVU test

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
Set/reset enable_ddl_from_pgendpoint GUC in upgrade scripts

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Set/reset GUC setting at required places

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
[Temporary Change] Use jira-babel-3135 engine branch for workflows

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
[Temporary change] Update engine branch for workflow

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
[Temp change] fix upd workflow

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
[Temp change] indent fix in workflow file

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
[Temp change] fix in workflow file

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Add exception for DDL statements while babelfish_dump_restore

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Add comment

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'babelfish-for-postgresql:BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
Merge branch 'babelfish-for-postgresql:BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
Add upgrade tests

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

KushaalShroff commented on contrib/babelfishpg_tsql/sql/information_schema_tsql.sql:
```diff
@@ -410,6 +410,43 @@ CREATE VIEW information_schema_tsql.table_constraints AS
+	FROM sys.pg_namespace_ext nc JOIN pg_class c ON (nc.oid = c.relnamespace)
+		LEFT OUTER JOIN sys.babelfish_namespace_ext ext ON nc.nspname = ext.nspname
+		LEFT OUTER JOIN sys.babelfish_view_def vd
+			ON ext.dbid = vd.dbid AND ext.orig_name = vd.schema_name AND c.relname = vd.object_name
```
We might need to add explicit collation for string comparison, this will start failing after Dipeshs changes go in

Deepesh125 replied:
Specifying explicit collation is no required here since we are joining sys objects only so collation conflict should not be occurred ideally. But it is always advisable and good practice to specify explicit collate clause.

HarshLunagariya replied:
Fixed

KushaalShroff commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -124,6 +130,10 @@ Datum init_catalog(PG_FUNCTION_ARGS)
 
+	/* bbf_view_def */
+	bbf_view_def_oid = get_relname_relid(bbf_view_def_TABLE_NAME, sys_schema_oid);
+	bbf_view_def_idx_oid = get_relname_relid(bbf_view_def_IDX_NAME, sys_schema_oid);
```
Let's keep the naming consistent, i.e. ALL CAPS

HarshLunagariya replied:
Fixed.

KushaalShroff commented on contrib/babelfishpg_tsql/src/dbcmds.c:
```diff
@@ -622,16 +624,21 @@ PG_FUNCTION_INFO_V1(create_builtin_dbs);
 	const char  *sa_name = text_to_cstring(PG_GETARG_TEXT_PP(0));
 
 	sql_dialect_value_old = GetConfigOption("babelfishpg_tsql.sql_dialect", true, true);
+	enable_ddl_from_pgendpoint_old = GetConfigOption(
```
If we have an extern variable pltsql_enable_ddl_from_pgendpoint in TSQL
Why do we need to use get/set config option?
IIRC then this is performance draining.

HarshLunagariya replied:
Got fixed indirectly, Now we're setting up GUC while initialising Babelfish.

KushaalShroff commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -1510,6 +1523,150 @@ modify_insert_stmt(InsertStmt *stmt, Oid relid)
+	 * Do not store definition/data in case of sys and information_schema_tsql
+	 * views
+	 */
+	if (strcmp(physical_schemaname, "sys") == 0 ||
```
We can re-use is_shared_schema function which is currently a static function

HarshLunagariya replied:
Done.

KushaalShroff commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -1510,6 +1523,150 @@ modify_insert_stmt(InsertStmt *stmt, Oid relid)
+	HeapTuple	tuple, reltup;
+	Form_pg_class	form_reltup;
+	int16		dbid;
+	char		*physical_schemaname, *logical_schemaname;
```
Free these pointers

HarshLunagariya replied:
Done.

KushaalShroff commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -1510,6 +1523,150 @@ modify_insert_stmt(InsertStmt *stmt, Oid relid)
+	HeapTuple	reltuple, scantup;
+	Form_pg_class	form;
+	int16		dbid;
+	char		*physical_schemaname, *logical_schemaname, *objectname;
```
Free these pointers

HarshLunagariya replied:
Done. No need to free objectname as it is referred from tuple.

KushaalShroff commented on test/JDBC/expected/13_4__preparation__bbf_view_def_upgrade-vu-prepare.out:
```diff
@@ -0,0 +1,13 @@
+~~ERROR (Message: relation "sys.babelfish_view_def" does not exist)~~
+
+
+CREATE VIEW dep_view_3135_2 AS SELECT * FROM information_schema.views;
```
Can we please add comments for this and what is expected (If fail, then why?)

HarshLunagariya replied:
Added.

HarshLunagariya replied:
Added.

KushaalShroff commented on contrib/babelfishpg_tsql/src/hooks.c:
```diff
@@ -1510,6 +1523,150 @@ modify_insert_stmt(InsertStmt *stmt, Oid relid)
+ * Stores view object's TSQL definition to bbf_view_def catalog
+ */
+static void
+pltsql_store_view_definition(const char *queryString, ObjectAddress address)
```
What happens when we run into error during this function? Will the view be created?

HarshLunagariya replied:
View won't be created and error will be thrown.

surendravishnoi commented on .github/workflows/minor-version-upgrade.yml:
```diff
@@ -4,12 +4,12 @@ on: [push, pull_request]
   extension-tests:
     env:
-      ENGINE_VER_FROM: BABEL_2_X_DEV_DEFAULT_mVU_SOURCE
+      ENGINE_VER_FROM: jira-babel-3135
```
It is working for others without any change, why did you need this? May be earlier it was an issue but not for a week or so

HarshLunagariya replied:
Fixed

surendravishnoi commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -875,6 +931,119 @@ get_user_for_database(const char *db_name)
+	const char	*logical_schema_name;
+	bool		is_tsql_view = false;
+
+	view_name = get_rel_name(relid);
```
who should free this memory?

HarshLunagariya replied:
It was unused function, missed to remove as part of refactoring. Removed now.

surendravishnoi commented on contrib/babelfishpg_tsql/src/catalog.c:
```diff
@@ -875,6 +931,119 @@ get_user_for_database(const char *db_name)
+		return false;
+	}
+	/* Fetch the relation */
+	bbf_view_def_rel = table_open(bbf_view_def_oid, RowExclusiveLock);
```
why exclusive lock?

HarshLunagariya replied:
It was unused function, missed to remove as part of refactoring. Removed now.

HarshLunagariya committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
Minor fixes

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Make is_shared_schema extern

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Add tsql dialect check in hook

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Free memory and change lock level for searching catalog

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Added comments in upgrade tests

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Minor fix

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Add collate clause for comparison

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'babelfish-for-postgresql:BABEL_2_X_DEV' into jira-babel-3135

surendravishnoi commented on contrib/babelfishpg_tsql/sql/object_definition_tsql.sql:
```diff
@@ -0,0 +1,14 @@
+	schema_name sys.SYSNAME NOT NULL,
+	object_name sys.SYSNAME NOT NULL,
+	definition sys.NTEXT,
+	is_ansi_nulls_on sys.BIT,
```
what if we need more flags in future? will be tricky to upgrade existing customer. Will it make more sense to have a 64 bit field so that we can cover up to 64 flags without an upgrade

HarshLunagariya replied:
Addressed introduced flag_validity(to track whether we're use corresponding bit in flag_values to represent any flag like ansi_nulls, quoted_identifier etc. related to view) and flag_values(to store actual values) int64 values in catalog.

surendravishnoi commented on test/JDBC/upgrade/13_4/preparation/bbf_view_def_upgrade-vu-prepare.sql:
```diff
@@ -0,0 +1,10 @@
+CREATE VIEW view_3135_upd AS SELECT 1;
```
Use filename_xxx as object name at all places to allow unique object names

HarshLunagariya replied:
Fixed

surendravishnoi commented on test/JDBC/upgrade/14_3/preparation/bbf_view_def_upgrade-vu-prepare.sql:
```diff
@@ -0,0 +1,10 @@
+CREATE VIEW view_3135_upd AS SELECT 1;
```
It seems all prepare files are same. If that is the case, add single file to input directory and add corresponding entry to schedule file

HarshLunagariya replied:
But output are different for them that's why I had to add in version specific folder. If I'll add it to schedule file then it will compare it's output with common expected file (i.e. bbf_view_def_upgrade-vu-prepare.out) and fail.

surendravishnoi commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--1.0.0--1.1.0.sql:
```diff
@@ -2,6 +2,8 @@
 
 SELECT set_config('search_path', 'sys, '||current_setting('search_path'), false);
+-- enable DDL from pgendpoint
+SELECT set_config('babelfishpg_tsql.enable_ddl_from_pgendpoint', 'true', false);
```
can we set it for whole create/update extension duration to avoid set/reset at multiple places

HarshLunagariya replied:
For create extension, We are setting/resetting at babelfishpg_tsql.in . But for for upgrade we've to add it to every script.

surendravishnoi commented on test/JDBC/upgrade/13_5/preparation/bbf_view_def_upgrade-vu-prepare.sql:
```diff
@@ -0,0 +1,10 @@
+CREATE VIEW view_3135_upd AS SELECT 1;
+GO
+
+CREATE VIEW dep_view_3135 AS SELECT * FROM sys.babelfish_view_def;
```
can we create a view on non-existent tables? this table is only available on 14 latest

HarshLunagariya replied:
No, it will fail. I will try to remove it in next iteration.

HarshLunagariya replied:
Fixed

HarshLunagariya committed:
Add cast for comparison using default collation

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Add cast to compare using collate

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Fix casts

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Use common 64 bit int to represent all the flags

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
remove unused check_is_tsql_view function

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'babelfish-for-postgresql:BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
Update minor-version-upgrade.yml

HarshLunagariya committed:
Merge branch 'babelfish-for-postgresql:BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
Rename objects in upgrade files

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Update expected file

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Empty Commit

HarshLunagariya committed:
Merge branch 'babelfish-for-postgresql:BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
Move verify script to 14_latest to latest dir

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Update expected output of verify scripts

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Unblock CREATE[OR REPLACE] VIEW in psql dialect from pg endpoint

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Add new testcases

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'babelfish-for-postgresql:BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
Cleanup code for set/reset enable_ddl_from_pgendpoint from sql scripts

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Revert "Cleanup code for set/reset enable_ddl_from_pgendpoint from sql scripts"

This reverts commit 096ca094dbbb742399bcd2c701ac05d85daf43bd.

HarshLunagariya committed:
Update expected_dependency.out to resolve failure

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Update tests

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'babelfish-for-postgresql:BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
empty commit

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
Revert "remove unused check_is_tsql_view function"

This reverts commit 393b3770a1890cd9fb0a8e2f7463ae7152fa0423.

HarshLunagariya committed:
Block ALTER/REPLACE stmts on TSQL views in PG dialect

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
minor fix

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
update test output

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Free memory in check_is_tsql_view

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Update/Add comments

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Update tests

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Rename enable_ddl_from_pgendpoint to enable_create_alter_view_from_pg

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Minor bug fix  and More test cases for GUC

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Fix tests

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Cleanup upgrade scripts and add check for shared schemas in check_is_tsql_view()

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Add funciton to get catalog oid

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Cleanup set/reset GUC from ownership.sql and babelfishpg_tsql.in

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Merge branch 'babelfish-for-postgresql:BABEL_2_X_DEV' into jira-babel-3135

HarshLunagariya committed:
Increase timeout for MVU Workflow

HarshLunagariya committed:
Revert "Increase timeout for MVU Workflow"

This reverts commit afe33eb04df9d2f050360e8d5fe0223261d2225c.

HarshLunagariya committed:
Revert temporary changes in workflow config

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>

HarshLunagariya committed:
Add LIKE clause in tests to make it deterministic

Signed-off-by: Harsh Lunagariya <lunharsh@amazon.com>
